### PR TITLE
rbd-mirror: Add sparse read for sync image

### PR DESCRIPTION
**Pull request by jazeltq**
Currently, the image sync do full read, and we shall add sparse read
to let the sync more efficiently.

Feature: http://tracker.ceph.com/issues/16780
Signed-off-by: tianqing <tianqing@unitedstack.com>

### CODE REVIEW

**Comment by jazeltq** [comment]
@dillaman

**Comment by dillaman** [comment]
@jazeltq unit tests are failing

**Comment by jazeltq** [comment]
@dillaman  ok, i will check....

File: src/tools/rbd_mirror/image_sync/ObjectCopyRequest.cc [hunk 1]
```diff
@@ -103,6 +103,8 @@ void ObjectCopyRequest<I>::send_read_object() {
 
   bool read_required = false;
   librados::ObjectReadOperation op;
+  std::map<uint64_t, uint64_t> extents;
+  bufferlist xbl;
```
Comment (dillaman) [inline src/tools/rbd_mirror/image_sync/ObjectCopyRequest.cc]: Local variable -- will be destructed before async operation completes. Update the tuple to provide storage for the extents and bufferlist.

File: src/tools/rbd_mirror/image_sync/ObjectCopyRequest.cc [hunk 2]
```diff
@@ -175,12 +184,26 @@ void ObjectCopyRequest<I>::send_write_object() {
...
   for (auto &sync_op : sync_ops) {
     switch (std::get<0>(sync_op)) {
     case SYNC_OP_TYPE_WRITE:
+      it = std::get<3>(sync_op).begin();
+      ::decode(extents, it);
+      ::decode(databl, it);
+      for(auto &it : extents){
+         bufferlist tmpbl;
+         len = it.second - it.first;
+         tmpbl.copy_in(begin, len, databl);
```
Comment (dillaman) [inline src/tools/rbd_mirror/image_sync/ObjectCopyRequest.cc]: Use bufferlist::substr_of to quickly build a new zero-copy write bl

File: src/tools/rbd_mirror/image_sync/ObjectCopyRequest.cc (hunk 1 above)
Comment (jazeltq) [inline src/tools/rbd_mirror/image_sync/ObjectCopyRequest.cc]: Ok. Thanks dillaman.

File: src/tools/rbd_mirror/image_sync/ObjectCopyRequest.cc (hunk 2 above)
Comment (jazeltq) [inline src/tools/rbd_mirror/image_sync/ObjectCopyRequest.cc]: Ok. Thanks dillaman.

File: src/test/librados_test_stub/TestMemIoCtxImpl.cc [hunk 3]
```diff
@@ -373,19 +372,29 @@ int TestMemIoCtxImpl::sparse_read(const std::string& oid, uint64_t off,
...
+      std::cerr << "  off: " << cit.get_start() << " len: " << cit.get_len() << std::endl;
+  }
+
   RWLock::RLocker l(file->lock);
   len = clip_io(off, len, file->data.length());
+
   if (m != NULL) {
     m->clear();
     if (len > 0) {
       (*m)[off] = len;
```
Comment (jazeltq) [inline src/test/librados_test_stub/TestMemIoCtxImpl.cc]: @dillaman  The sparse_read here is not right. It just put the length to the extents. How to use bufferlist to emulate sparse_read here ?

Comment (dillaman) [inline src/test/librados_test_stub/TestMemIoCtxImpl.cc]: For the unit tests, it's probably safe enough to just populate the entire requested extent even if it doesn't exist. This test stub doesn't track holes, so you wouldn't get a true implementation.

**Comment by jazeltq** [comment]
ping @dillaman

File: src/tools/rbd_mirror/image_sync/ObjectCopyRequest.cc [hunk 4]
```diff
@@ -173,14 +172,19 @@ void ObjectCopyRequest<I>::send_write_object() {
 
   auto &sync_ops = m_snap_sync_ops.begin()->second;
   assert(!sync_ops.empty());
-
+  uint64_t begin;
```
Comment (dillaman) [inline src/tools/rbd_mirror/image_sync/ObjectCopyRequest.cc]: Minor: can you use a more descriptive variable name? Perhaps something like ```buffer_offset```

**Comment by jazeltq** [comment]
ping @dillaman

**Pull request merged**